### ISSUE TITLE
Deferred chain resolution breaking change

### ISSUE THREAD

**Issue by Krinkle** [jquery/jquery#3646]
### Description ###

The way a chained promise is resolved in jQuery 3 seems to have changed compared to how it was in jQuery 1.x and 2.x. This does not appear to be related to the async behaviour change, which is well documented.

I have not yet been able to fully pinpoint the problem, but the end result is that jQuery is not consistently resolving a promise with the resolution value. Instead, if the resolution value is itself a promise, it can decide to hide this promise from the caller and instead waits for it directly.

This is undesirable as not all promise chains are a single continuous actions. Having a handler in between the different actions is mandatory and expected.

At Wikimedia, the "OOjs UI" library has an interface for confirmation dialogs. It works like this:

```js
OO.ui.confirm( /* string */ text ).then( function ( /* boolean */ result ) {
 /* .. */
} );
```

The returned promise resolves with a boolean value true/false depending on which button the user pressed.

Internally this method uses a promise to create a dialog, wait for its opening animation, and wait for its closing/cleanup actions.

Source code: <https://github.com/wikimedia/oojs-ui/blob/v0.21.1/src/windows.js#L52-L85>
```js
OO.ui.confirm = function ( text, options ) {
	return OO.ui.getWindowManager().openWindow( 'message', $.extend( {
		message: text
	}, options ) ).then( function ( opened ) {
		return opened.then( function ( closing ) {
			return closing.then( function ( data ) {
				return $.Deferred().resolve( !!( data && data.action === 'accept' ) );
			} );
		} );
	} );
};
```

Most promises return a primitive value or an object with methods that can be called to yield another promise. However in this case, the promise is actually resolved with another promise. The resolution indicates that the opening is done (`opening.then`) and the value given is the promise that will be resolved once the dialog is closed. The latter is malfunctioning in jQuery 3.

It appears it is internally unwrapping this promise layer, thus causing a mismatch between the actual and expected value.

Downstream bug report: <https://phabricator.wikimedia.org/T163510>

### Link to test case ###

Minimal test case at: <https://codepen.io/Krinkle/pen/JNbNzg?editors=0010>

Compare jQuery 1.11.3, 1.12.4, 2.2.4 and 3.2.1 - with only the latter failing.

Problems highlighted in red:

<img width="619" alt="screen shot 2017-04-25 at 13 02 52" src="https://cloud.githubusercontent.com/assets/156867/25405233/827824bc-29b7-11e7-9234-265273412171.png">

I'd be fine with working around it if this is considered an invalid use case. However then we should make sure that this change  is well-documented and that the suggested alternative way to do this works both in jQuery 1.x and jQuery 3.x.

**Comment by mgol** [comment]
This thenable-uwrapping is required by the Promises/A+ spec; specifically [the promise resolution algorithm](https://promisesaplus.com/#the-promise-resolution-procedure). The relevant point is [2.3.3.3](https://promisesaplus.com/#point-56) which describes thenable unwrapping. In particular, [point 2.3.3.3.1](https://promisesaplus.com/#point-57) mandatest that if a thenable "resolves" to a promise, the whole algorithm is repeated until you get something that is not a thenable.

So it seems it's purely a documentation issue?

**Comment by mgol** [comment]
I'm a little confused about this message, though:

> jQuery.Deferred exception: opened.then is not a function

If an object doesn't have a `then` method then it's not a thenable and it needs to be treated as a regular value and fulfill the promise with it if that's what you get in a `then` handler. I'm not sure how we can get such an error then unless there's some bug in Deferreds.

**Comment by Krinkle** [comment]
@mgol Consider the following simplified example, involving `$.when()` to trigger the bug:

```js
// Enable warnings
jQuery.Deferred.exceptionHook = function( error, stack ) {
 console.warn( "jQuery.Deferred exception: " + error.message, error.stack, stack );
};

function System() {
  var shutdown = $.Deferred();
  this.start = function() {
    return $.Deferred(function (d) {
      setTimeout(function () {
        d.resolve(shutdown);
      }, 100);
    });
  };
  this.stop = function (reason) {
    /* do stuff.. */
    shutdown.resolve(reason);
  };
}

// Works
// > Hello world!
// > Bye!, reason: Meh!
var x = new System();
x.start().then(function (shutdown) {
  console.log('Hello world!');
  return shutdown.then(function (reason) {
    console.log('Bye!, reason:', reason);
  });
});

setTimeout(function () {
  x.stop('Meh!');
}, 200);


// Does not work
// > Hello world!
// > jQuery.Deferred exception: shutdown.then is not a function
var x = new System();
$.when(x.start()).then(function (shutdown) {
  console.log('Hello world!');
  return shutdown.then(function (reason) {
    console.log('Bye!, reason:', reason);
  });
});

setTimeout(function () {
  x.stop('Meh!');
}, 200);
```

**Comment by dmethvin** [comment]
Right, so you were thinking it resolved as a Promise with a `.then` method but it didn't because it was unwrapped.

**Comment by Krinkle** [comment]
Yeah, but it seems odd that it only happens in `$.when()` and when there are other levels of indirection with return values to `.then()`, but not when passed to `resolve()` - as in the first example.

It's unclear to me whether that should also be expected behaviour.

**Comment by gibson042** [comment]
@Krinkle you're smack in the middle of the legacy–standard interface here, which is why behavior is odd. The crux of it is that "the promise is actually resolved with another promise" has a counterintuitive interpretation when dealing with standard promises—["resolved" is distinct from "settled"](http://ecma-international.org/ecma-262/7.0/#sec-promise-objects), and thenables can never be fulfillment values (and thus can never be provided to `.then` onFulfilled callbacks) because [resolution](http://ecma-international.org/ecma-262/7.0/#sec-promise-resolve-functions) intercepts them and enqueues a [PromiseResolveThenableJob](http://ecma-international.org/ecma-262/7.0/#sec-promiseresolvethenablejob) instead of fulfilling the promise. However, jQuery Deferreds have no such restriction (i.e., we _do_ allow fulfillment with thenables, and therefore don't need the concept of fulfillment as separate from resolution), and backwards compatibility motivated us to continue supporting
```js
jQuery.Deferred().resolve( thenable )
	.then( val => console.assert(val === thenable) )
```
even while standards compliance _requires_
```js
jQuery.Deferred().resolve()
	.then( () => jQuery.Deferred().resolve("foo") )
	.then( val => console.assert(val === "foo") )
```
...which brings us to `jQuery.when( thenable )`, which is analogous to `Promise.resolve( thenable )` and unwraps values as necessary.

In short, it is dangerous to use `.then` and/or `jQuery.when` with thenable-fulfilled Deferreds. You should either wrap the thenables (e.g., in arrays or non-thenable objects), or update the callbacks to handle non-thenable arguments (e.g., `jQuery.when( closing ).then` instead of `closing.then`).

Our team discussion also raised the need for improved documentation... https://github.com/jquery/api.jquery.com/issues/921 seems to cover part of it, but I'm thinking that it's insufficient on its own. Is there any other place some text could have appeared that would have clarified behavior for you?

**Comment by Krinkle** [comment]
@gibson042 I suppose it'd have been nice if https://api.jquery.com/jQuery.Deferred/, <https://api.jquery.com/deferred.resolve/> and/or <https://api.jquery.com/deferred.then/> clearly state what the result is of providing a Thenable to resolve() or as return value to `then()`.

Afaik doing that (returning or passing Thenable in those cases) is supported and standards-compliant. The only question is what it will do.

In ES6 (tested using latest Chrome), it does:

```js
a = Promise.resolve(123)
// Promise {[[PromiseStatus]]: "resolved", [[PromiseValue]]: 123}

b = Promise.resolve(a)
// Promise {[[PromiseStatus]]: "resolved", [[PromiseValue]]: 123}

c = new Promise(function (resolve) { resolve(a); });
// Promise {[[PromiseStatus]]: "resolved", [[PromiseValue]]: 123}

d = Promise.resolve(400).then( () => { return b; })
// Promise {[[PromiseStatus]]: "resolved", [[PromiseValue]]: 123}
```

In jQuery 1.x, A and D behave the same. B behaving differently kind of makes sense to me for back-compat, and also because if in such case you knowingly have a Promise, one just as well use `a` directly, or `a.then()` perhaps.

The problem in this case happened because of C, which is imho are fair situation to be in, e.g. manually promisifying some async process and being given back a Thenable value from one of your internal async steps. (This may even happen as part of an upstream change, e.g. something used to be sync but now becomes async via a promise.)

I understand that for jQuery internally B and C are the same. But I'd say it's a lot more likely that B happens and wants back-compat no-unwrapping, than situation C, as C feels more likely to happen when calling other code that may return a Promise and is expected to be unwrapped.

```js
a = $.Deferred().resolve(123)
a.done( val => console.log(val) )
// 123

b = $.Deferred().resolve(a)
b.done( val => console.log(val) )
// {resolve: ƒ, resolveWith: ƒ, reject: ƒ, rejectWith: ƒ, notify: ƒ, …}

c = $.Deferred(function (df) { df.resolve(a) });
c.done( val => console.log(val) )
// {resolve: ƒ, resolveWith: ƒ, reject: ƒ, rejectWith: ƒ, notify: ƒ, …}

d = $.Deferred().resolve(400).then( () => { return a; });
d.done( val => console.log(val) )
// 123
```

I feel that when calling `d.resolve()` it's a lot like returning from inside a chain of `then()` callbacks. You just provide the next value that may or may not also be a Promise to be waited for. I suppose it could be worked around by using `a.then(df.resolve, df.reject)`.

Either way, aside from explicit documentation (for new comers) that jQuery now does unwrapping, it would be good to also point out where jQuery specifically doesn't unwrap promises.

And perhaps we can deprecate this and fix as breaking change in jQuery 4?

**Comment by gibson042** [comment]
Thanks, I moved the documentation needs to https://github.com/jquery/api.jquery.com/issues/1039 . And I guess I'll keep this issue open until we either always unwrap thenables, or decide not to make that change.

**Comment by cherish007** [comment]
pretty

**Comment by timmywil** [comment]
I think we all agree that we should be unwrapping to match standard Promise behavior, but it'll have to wait until 4.0 since it's a breaking change.
